pkg/kudoctl/kudoinit/setup/setup.go

```diff
@@ -39,6 +45,79 @@ func NewInstaller(options kudoinit.Options, crdOnly bool) *Installer {
 	}
 }
 
+// Validate checks that the current KUDO installation is correct
+func (i *Installer) VerifyInstallation(client *kube.Client, result *verifier.Result) error {
+	// Check if all steps are correctly installed
+	for _, initStep := range i.steps {
+		if err := initStep.VerifyInstallation(client, result); err != nil {
+			return fmt.Errorf("error while verifying init step %s: %v", initStep.String(), err)
+		}
+	}
+
+	return nil
+}
+
+func requiredMigrations() []migration.Migrater {
+
+	// Determine which migrations to run
+	return []migration.Migrater{
+		// Implement actual migrations
+	}
+}
+
+func (i *Installer) PreUpgradeVerify(client *kube.Client, result *verifier.Result) error {
+	// Step 1 - Verify that upgrade can be done
+	// Check if all steps are upgradeable
+	for _, initStep := range i.steps {
+		if err := initStep.PreUpgradeVerify(client, result); err != nil {
+			return fmt.Errorf("error while verifying upgrade step %s: %v", initStep.String(), err)
+		}
+	}
+	if !result.IsValid() {
+		return nil
+	}
+
+	// Step 2 - Verify that any migration is possible
+	migrations := requiredMigrations()
+	clog.Printf("Verify that %d required migrations can be applied", len(migrations))
+	for _, m := range migrations {
+		if err := m.CanMigrate(client); err != nil {
+			result.AddErrors(fmt.Errorf("migration %s failed install check: %v", m, err).Error())
+		}
+	}
+
+	// TODO: Verify existing operators and instances?
+
+	return nil
+}
+
+// Upgrade an existing KUDO installation
+func (i *Installer) Upgrade(client *kube.Client) error {
```

Comment:
Could we use a Patch instead of deleting/re-installing:

https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/apps/v1/statefulset.go#L50

Reply:
We could - We actually patch the CRDs, but for the Service and statefulset this would involve either manually creating the patch, or saving the old version in annotations, similar to TaskApply. Delete/Recreate is a lot easier. We may come back to patching later on though.